### DISCUSSION TITLE
Running background jobs

### DISCUSSION THREAD

**Discussion by tavindev** [vercel/next.js#33989] [Help]
Is there any serverless solution for scheduling background jobs? I need to track a 3rd party transaction/trade and check if it is completed within 5 minutes. I came up with other ideas, but none of them seem as good as background jobs.

I have thought about making the client periodically call a serverless function that checks if the transaction/trade was completed, but it's a very unreliable approach. If the client closes the browser tab or gets disconnected, the tracking stops, so that's not a good idea.

Another approach would be to make a recursive serverless function. I'm not sure if this would work and it seems like a bad idea.

Is there any other solution or should I make a separate server that is responsible for tracking 3rd party transactions/trades and updating the database accordingly?

**Comment by juanbzz** [comment]
What you need is a serverless task queue service. I was pretty sure Vercel had one in their integrations, but I can't find it anymore.

I think AWS SQS is what you want, but you can see a list of alternatives in the [Task Queues website](https://taskqueues.com/).

**Reply by tavindev** [reply]
Great, thanks for your suggestion. I've never heard of AWS SQS, so I did a little research on it. Based on what I've found, I assume the architecture should be the following: as the 3rd party transaction/trade gets created, we add a message containing the transaction id to the queue. SQS will then trigger a Lambda Function that receives the id and tracks the transaction. Am I right?

**Comment by AlessandroVol23** [comment]
You need to decouple both processes with a message queue. Like @j0lv3r4 already suggested AWS SQS is a great option for that.

>  need to track a 3rd party transaction/trade and check if it is completed within 5 minutes.

How do you check that? Are you checking some DB entries or calling some REST API to check if it is finished?

Using webhooks would be perfect for that but that depends on the third party. Then it would simply call your API when the third party is finished. This will probably not be supported so let's look at another way with a message queue.

>  as the 3rd party transaction/trade gets created, we add a message containing the transaction id to the queue. SQS will then trigger a Lambda Function that receives the id and tracks the transaction. Am I right?

Yes, kind of. You've got two more issues I think:

1. Lambda timeout
2. You don't know when it will be finished

So you need to regularly call a lambda function to check if it is finished and can't leave it running in a loop (this is an anti-pattern and expensive). 

For that, you would also need a cron-job to call your lambda function every 5 minutes for example. For the cron-job, you could use something like GitHub actions.

For the message queue, I'd recommend (shamelessly) the service I've built: https://serverlessq.com 
You can simply reuse your existing API routes within Vercel and send your messages to your message queue. I am adding support for Cron Jobs pretty soon as well.
With that, you don't need to add another huge component (i.e. AWS) to your application and you don't need to learn all about SQS. 

Let me know if you have any other questions I hope that helps 🙂

**Reply by deadcoder0904** [reply]
@AlessandroVol23 serverlessq is down. is there an easier way to run with next.js + docker? i want to do it without using a 3rd-party service as bullmq would work well for my needs.

**Reply by maniraj1234** [reply]
@deadcoder0904 , I also wanted same. so I used a nodejs background worker with a docker and scheduled it using container app job in azure.

**Comment by mixdev** [comment]
There are some ways to run workers in NextJS. Like this -> https://gist.github.com/babldev/8c7b8441abcab7526786a78a73b7bebd SO you may manage communication between your web application and the workers through messagequeue/ pubsub (BullMQ/RSMQ etc.,)
However, you would want to manage loadbalancing / reliability etc yourself. 

Another option is to offload this to an external service (like what many have mentioned above). However there are limitations on that as well.

**Comment by prageeth** [comment]
A bit late to the party, but for anyone stumbling on this post, checkout: https://dispatched.dev/
